### ISSUE TITLE
Memoize subscription aggregators

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/action#196]
Memoizing a subscription aggregator brings about a really interesting Problem. Let's say we foolishly tried to do it like this:

```
const makeProjectSubs = (teamMembers) => {
  if (teamMembers !== makeProjectSubs.teamMembers) {
    makeProjectSubs.teamMembers = teamMembers;
    const projectSubs = [];
    for (let i = 0; i < teamMembers.length; i++) {
      const teamMemberId = teamMembers[i].id;
      projectSubs[i] = cashay.subscribe(projectSubString, subscriber, {
        op: 'projectSub',
        key: teamMemberId,
        variables: {teamMemberId}
      });
    }
    makeProjectSubs.cache = projectSubs;
  }
  return makeProjectSubs.cache;
};
```

This won't return any results because `projectSubs[i]` points to the initial object created by cashay. Since that time, cashay overwrote this old object with a completely new one. It does this for immutability reasons: by making `cashay.subscribe` return a new object whenever the underlying data is changed, react-redux knows that it should call render again. If we just mutated the underlying data, then we'd never rerender.

So we're at a catch 22-- if we mutate the current object, then react-redux assumes it's the same old stuff & will set `shouldComponentUpdate == false`. If we don't mutate the current object, we memoize ancient data. 

I've been hesitant to do it, but I think it might be best if redux didn't listen to cashay changes. The timing is pretty convenient for this to come out: https://github.com/reactjs/react-redux/pull/416, so I'll play around and see what happens.

**Comment by jordanh** [comment]
That react-redux@next thread was a very inspirational read. _Lots_ of interesting options there via `connectAdvanced`.

Just for funsies and completeness, I wanted to explore a few other options that came to mind. Consider them thought starters.

## Returning a function

Instead of a Cashay query returning an object, what if it optionally returned a function? Then you could memoize like this:

```javascript
const makeProjectSubs = (teamMembers) => {
  if (teamMembers !== makeProjectSubs.teamMembers) {
    makeProjectSubs.teamMembers = teamMembers;
    const projectSubs = [];
    for (let i = 0; i < teamMembers.length; i++) {
      const teamMemberId = teamMembers[i].id;
      projectSubs[i] = cashay.subscribe(projectSubString, subscriber, {
        op: 'projectSub',
        key: teamMemberId,
        variables: {teamMemberId},
        returnFunction: true
      });
    }
    makeProjectSubs.cache = projectSubs;
  }
  return makeProjectSubs.cache;
};
```

...and mapStateToProps would look like this:

```javascript
const mapStateToProps = () => ({ // not shown, teamMembers
  projectSubs: makeProjectSubs(teamMembers).map((ps) => ps())
});
```

**Pros:** simple, uses existing interface, doesn't require anything from the future
**Cons:** still O(n), ugly


## Creating a Cashay memoization interface

Cashay could provide its own memoization interface. Here's one in three functions:

   * `cashay.memoizationCreate()`: returns an opaque handle to a memoization query
   * `cashay.memoizationDestroy(handle)`: removes an old memoization handle
   * `cashay.memoizedQueries(handle, queries)`: given an opaque handle and an array of Cashay queries, it returns a new object if and only if one of the queries have new data.

It look like this:

```javascript
const makeProjectSubs = (teamMembers) => {
  if (teamMembers !== makeProjectSubs.teamMembers) {
    makeProjectSubs.teamMembers = teamMembers;
    const projectSubs = [];
    for (let i = 0; i < teamMembers.length; i++) {
      const teamMemberId = teamMembers[i].id;
      projectSubs[i] = cashay.subscribe(projectSubString, subscriber, {
        op: 'projectSub',
        key: teamMemberId,
        variables: {teamMemberId},
        returnFunction: true
      });
    }
    if (makeProjectSubs.cache) {
      cashay.memoizationDestroy(makeProjectSubs.cache);
    } 
    makeProjectSubs.cache = [
      cashay.memoizationCreate(),
      projectSubs,
    ];
  }
  return makeProjectSubs.cache;
};

const mapStateToProps = () => ({ // not shown, teamMembers
  projectSubs: cashay.memoizedQueries(...makeProjectSubs(teamMembers))
});
```

Implementation would be fairly straightforward:

1. If array object given to `memoizedQueries` changes:
   1. Update invalidation dependencies, return new results
1. If array object given to `memoizedQueries` does not change:
   1. If no new query data indicated by invalidations, return cached results
   2. If new query data indicated by invalidations, return new results
1. When new data arrives:
   1. Use memoization invalidation dependencies to invalidate affected queries

**Pros:** O(1), uses existing connect pattern
**Cons:** adds additional cashay calls

**Comment by mattkrick** [comment]
I think the second idea is headed in the right direction! Basically, we need to handle both the invalidation and caching within cashay in a pattern that works for all computed values. The pattern shown is pretty similar to MDG's Tracker, or more recently MobX. Couple questions:
- Should we support joining different subs? Creating new fields? For example, what we do in the `createMembers` duck. 
- Should we support joining subs & queries?
- Should we support joining subs and local data?

I wanna say yes, support it all.
- Since the computed data fields may not exist on the server, we can't normalize it, so it doesn't belong in our store. Rather it'd be in something like `cashay.cachedComputations`.
- Figuring out what subs/queries/local data we depend on is tricky without using the magic of proxies
- The current cashay dependency system works off of `entities`, there is no dependency tracking on `result` (the root call, in this case, the array of IDs that compose the subscription).

Working off of your second example, here's a rough API:
```
const projectSubs = cashay.computed('projectSubs', {
  // these are variables that are outide of the cashay ecosystem. If they change, we invalidate the cached result
  variables: {
    tms: state.auth.obj.tms
  },
  // The result of this function is cached at `cashay.cachedComputations.projectSubs[key='']`
  resolve: ({tms}, dependency) => {
    const teamMemberIds = tms.map(teamId => `${userId}::${teamId}`);
    const projectSubs = [];
    for (let i = 0; i < teamMemberIds.length; i++) {
      const teamMemberId = teamMemberIds[i];
      projectSubs[i] = cashay.subscribe(projectSubString, subscriber, {
        op: 'projectSub',
        key: teamMemberId,
        variables: {teamMemberId},
        // Whenever cashay.cachcedSubscriptions.projectSub[teamMemberId] is invalidated, we also invalidate the computation
        dependency
      });
    }
    return projectSubs;
  }
})
```

**Comment by jordanh** [comment]
That is a good, simple interface. I really like it. Let's try it!

Alternatively, I could see this maybe being it's own module. Right?

**Comment by mattkrick** [comment]
Here's a taste: 
```
const resolve = (teamId) => {
  const presenceSubOptions = {
    variables: {teamId},
    op: 'socketWithPresence',
    dependency: 'editingPresence'
  };
  const presenceSub = cashay.subscribe(presenceSubscription.string, presenceSubscriber, presenceSubOptions);
  const editing = presenceEditingHelper(presenceSub.data.presence);
  return {
    presenceSub: {
      ...presenceSub,
      data: {
        ...presenceSub.data,
        editing
      }
    }
  };
};

const mapStateToProps = (state, props) => {
  return cashay.computed('editingPresence', [props.params.teamId], resolve)
};
```

Questions:
- `computed`: Is this the right name? This is what MobX uses, but I like my methods to be verbs.
- `dependency`: I opted for a string rather than passing a function down & introducing another closure. Is this the right name? Maybe computation? computed? Something to make the link between the 2 stronger?
- API: `cashay.computed(op, inputs, resolve)`. Currently, `inputs` only takes an array. I do this because the function will realistically get called hundreds of thousands of times & `Array.isArray` isn't free. But internally we could convert a single item into an array. Any preference?

**Comment by mattkrick** [comment]
Oh! And worth noting that the resolve function args look like `resolve(...inputs)`. It's a _little_ magical, but no more than `apply` would be.

**Comment by mattkrick** [comment]
Last doozie of a question: should computed values be able to depend on other computed values?
I wanna say no, and force folks to just use `cashay.compute` inside the other computed value, otherwise it kinda feels like we're stacking turtles...

**Comment by jordanh** [comment]
Ok, MHOs:

- Really love the interface overall
- `cashay.computed` is a very good name and args `(op, inputs, resolve)` are great
- For `dependency` how about `computationDependency` it's longish (but so is `mutationHandler` kinda) but it tells you what it does and how it relates
- For should computer values be able to depend on other computed values... for now let's say no. Success would look like somebody asking for this and then explaining WHY WHY WHY. We'll learn something then!

This looks _great_!

**Comment by mattkrick** [comment]
Fixed cashay@0.18.1
